src/lib/cloud-provider.js

```diff
@@ -0,0 +1,124 @@
+import log from './log.js';
+
+
+class CloudProvider {
+    constructor (cloudHost, vm, user, projectId) {
+        this.vm = vm;
+        this.username = user;
+        this.projectId = projectId;
+
+        // Open a websocket connection to the clouddata server
+        this.openConnection(cloudHost);
+    }
+
+    /**
+     * Open a new websocket connection to the clouddata server.
+     * @param {string} cloudHost The cloud data server to connect to.
+     */
+    openConnection (cloudHost) {
+        if (window.WebSocket === null) {
+            log.warn('Websocket support is not available in this browser');
+            this.connection = null;
+            return;
+        }
+
+        this.connection = new WebSocket((location.protocol === 'http:' ? 'ws://' : 'wss://') + cloudHost);
+
+        this.connection.onerror = e => {
+            log.error(`Websocket connection error: ${JSON.stringify(e)}`);
+
+            // TODO Add re-connection attempt logic here
+        };
+
+        this.connection.onmessage = event => {
+            const messageString = event.data;
+            log.info(`Received websocket message: ${messageString}`);
+            const message = JSON.parse(messageString);
+            if (message.method === 'set') {
+                const varData = {
+                    projectId: this.projectId,
+                    varUpdate: {
+                        projectId: message.project_id,
+                        name: message.name,
+                        value: message.value
+                    }
+                };
+                this.vm.postIOData('cloud', varData);
+            }
+        };
+
+        this.connection.onopen = () => {
+            this.writeToServer('handshake');
+            log.info(`Successfully connected to clouddata server.`);
+        };
+
+        this.connection.onclose = () => {
+            log.info(`Closed connection to websocket`);
+        };
+    }
+
+    /**
+     * Format and send a message to the cloud data server.
+     * @param {string} methodName The message method, indicating the action to perform.
+     * @param {string} dataName The name of the cloud variable this message pertains to
+     * @param {string | number} dataValue The value to set the cloud variable to
+     * @param {number} dataIndex The index of the item to update (for cloud lists)
+     * @param {string} dataNewName The new name for the cloud variable (if renaming)
+     */
+    writeToServer (methodName, dataName, dataValue, dataIndex, dataNewName) {
+        const msg = {};
+        msg.method = methodName;
+        msg.user = this.username;
+        msg.project_id = this.projectId;
```

Comment:
Should this require some form of authentication? (I.e. like `?x-token`.) The absolute last thing anyone wants is for a user to impersonate some other user in the cloud data log by overriding `msg.user` to be something other than their own username.

Reply:
@towerofnix This looks the same as 2.0 cloud protocol; where one uses cookies (in the upgrade request, I think) to confirm authenticity.

Reply:
@towerofnix, the messages implicitly use session data, and authentication on the server fails without it.